src/libraries/System.Net.Http/src/System/Net/Http/DiagnosticsHandler.cs

```diff
@@ -13,36 +13,64 @@ namespace System.Net.Http
     /// <summary>
     /// DiagnosticHandler notifies DiagnosticSource subscribers about outgoing Http requests
     /// </summary>
-    internal sealed class DiagnosticsHandler : DelegatingHandler
+    internal sealed class DiagnosticsHandler : HttpMessageHandlerStage
     {
         private static readonly DiagnosticListener s_diagnosticListener =
                 new DiagnosticListener(DiagnosticsHandlerLoggingStrings.DiagnosticListenerName);
 
-        /// <summary>
-        /// DiagnosticHandler constructor
-        /// </summary>
-        /// <param name="innerHandler">Inner handler: Windows or Unix implementation of HttpMessageHandler.
-        /// Note that DiagnosticHandler is the latest in the pipeline </param>
-        public DiagnosticsHandler(HttpMessageHandler innerHandler) : base(innerHandler)
+        private readonly HttpMessageHandler _innerHandler;
+        private readonly TextMapPropagator _propagator;
+        private readonly HeaderDescriptor[]? _propagatorFields;
+
+        public DiagnosticsHandler(HttpMessageHandler innerHandler, TextMapPropagator? propagator = null, bool autoRedirect = false)
         {
+            Debug.Assert(IsGloballyEnabled());
+
+            _innerHandler = innerHandler;
+            _propagator = propagator ?? TextMapPropagator.Current;
+
+            if (autoRedirect)
+            {
+                // Prepare HeaderDescriptors for fields we need to clear when following redirects
+                // If a PassThroughPropagator is used, we don't have to clear the headers as they will always be the same
+                if (_propagator.Fields is IReadOnlyCollection<string> fields &&
+                    fields.Count > 0 &&
+                    !ReferenceEquals(_propagator, TextMapPropagator.CreatePassThroughPropagator()))
```

Comment:
>!ReferenceEquals(_propagator, TextMapPropagator.CreatePassThroughPropagator() [](http://example.com/codeflow?start=20&length=77)

now I know why you added the test for checking the returned instances :-)

Reply:
I am wondering is it worth to optimize for this case?

Reply:
It's not a must-have, but I don't see why not do it if it's just this one check. It saves a few allocations in the ctor and some cycles (5 dictionary removes) on every redirect.

Are you concerned because it's making an assumption of reference equality?

Reply:
> Are you concerned because it's making an assumption of reference equality?

Yes.

Reply:
> I don't see why not do it if it's just this one check

If this may not be idempotent, then this check will actually add overhead, as we'll allocate the additional propagator that will never match and then still have to do all the subsequent work.  Something to consider.

Reply:
I can add an assert here that would catch if we ever change the behavior to return different instances (I added a test for this in Diagnostics tests already).

This optimization isn't incredibly important, but we may choose to make similar ones in the future. For example, we could avoid allocating an `Activity` per request when `NoOutputPropagator` is used and there are no other listeners present.

I think the broader question here is whether we should be relying on implementation details like this in libraries. We already do (e.g. Roslyn's tratment of `static ROS<byte> => new byte[]` / when making decisions on which things to cache), but is that more of an exception to the rule?

Reply:
> This optimization isn't incredibly important

I would remove it in that case : )

> I think the broader question here is whether we should be relying on implementation details like this in libraries

I would evaluate these on a case-by-case basis. If they are important optimizations I don't mind commiting to them but DiagnosticSource has the extra challenge that users can upgrade their dependency to any future version of it via the out-of-band NuGet package. So any dependency this code takes on its implementation becomes a requirement that every future version of the DiagnosticSource library must maintain.

Reply:
Ah right it's shipping OOB :/

I'll revert this part